# Exercise 2.A Fundamentals of pandas

# Exercise 2.A Fundamentals of pandas

In [ ]:
import numpy as np
import pandas as pd #Had issues importing pandas, relied on Copilot to assist me in correcting error


ModuleNotFoundError: No module named 'pandas'

In [5]:
import sys
!{sys.executable} -m ensurepip --default-pip

Looking in links: c:\Users\mruiz\AppData\Local\Temp\tmpwfdudamw


In [6]:
!{sys.executable} -m pip install pandas

   ---------------------------------------- 0.0/9.8 MB ? eta -:--:--
   ------- -------------------------------- 1.8/9.8 MB 9.1 MB/s eta 0:00:01
   --------------------- ------------------ 5.2/9.8 MB 11.5 MB/s eta 0:00:01
   --------------------------------- ------ 8.1/9.8 MB 11.7 MB/s eta 0:00:01
   ---------------------------------------- 9.8/9.8 MB 11.5 MB/s eta 0:00:00



[notice] A new release of pip is available: 25.0.1 -> 26.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [7]:
import pandas as pd

## Series.

In [44]:
aa = [1,2,3,4] #list
ab = (5,6,7,8) #tuple
ac = {9,10,11,12} #set
ad = {'first': 13,'second': 14,'third': 15,'fourth': 16} #dictionary

In [45]:
ser1 = pd.Series(aa)
print(ser1)
ser2 = pd.Series(ab)
print(ser2)
###ser3 = pd.Series(ac) #Gives error that says "type is unordered" and this is not an ordered seq
ser3 = pd.Series(list(ac)) #Had to change this to a list in order to complete "DataFrames from Series."
print(ser3)
ser4 = pd.Series(ad)
print(ser4)

0    1
1    2
2    3
3    4
dtype: int64
0    5
1    6
2    7
3    8
dtype: int64
0     9
1    10
2    11
3    12
dtype: int64
first     13
second    14
third     15
fourth    16
dtype: int64


In [46]:
ser5 = pd.Series(17, 18, 19, 20) #TypeError: Index(...) must be called with a collection of some kind, 18 was passed

TypeError: Index(...) must be called with a collection of some kind, 18 was passed

## DataFrames from Series.

In [53]:
df1 = pd.DataFrame(ser1)
df1

,0
0,1
1,2
2,3
3,4


In [54]:
#df1.insert(1, "column2", ser2)
#df1.insert(1, "column3", ser3) # I kept getting error NameError: name 'ser3' is not defined
#df1 #no matter what I did it was not running, even when I went up and reran all the ser 

df1.insert(1, "column2", ser2)
df1.insert(2, "column3", ser3)
df1

,0,column2,column3
0,1,5,9
1,2,6,10
2,3,7,11
3,4,8,12


In [55]:
df1.insert(2,'column4', ser4)
df1

,0,column2,column4,column3
0,1,5,NaN,9
1,2,6,NaN,10
2,3,7,NaN,11
3,4,8,NaN,12


In [56]:
df1.loc[4] = [5, 9, 12, 0]
df1


,0,column2,column4,column3
0,1,5,NaN,9
1,2,6,NaN,10
2,3,7,NaN,11
3,4,8,NaN,12
4,5,9,12.0,0


## Reading Data into a DataFrame.

Create a new variable df2. Use the read_csv function to read the contents of sakila_example.csv into df2. (Note that you will need to save the CSV file to the same folder where Jupyter Notebook is running for this exercise.) Call df2 to review the result.

In [58]:
df2 = pd.read_csv('sakila_example.csv')
df2

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE
...,...,...,...,...,...
257,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,GARY PENN
258,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,CHRISTOPHER WEST
259,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,DUSTIN TAUTOU
260,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,ANGELA HUDSON


Display the top 5 records from df2.

In [59]:
df2.head(5)

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE


Display the last 10 records from df2.

In [60]:
df2.tail(10)

,title,description,category,language,actor_name
252,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,MILLA KEITEL
253,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,GROUCHO DUNST
254,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,BURT TEMPLE
255,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,UMA WOOD
256,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,RIP WINSLET
257,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,GARY PENN
258,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,CHRISTOPHER WEST
259,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,DUSTIN TAUTOU
260,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,ANGELA HUDSON
261,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,JAMES PITT


Show summary information about df2 using the info method.

In [61]:
df2.info()

<class 'pandas.DataFrame'>
RangeIndex: 262 entries, 0 to 261
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   title        262 non-null    str  
 1   description  262 non-null    str  
 2   category     262 non-null    str  
 3   language     262 non-null    str  
 4   actor_name   262 non-null    str  
dtypes: str(5)
memory usage: 10.4 KB


Based on the output I can determine that there are a total of 262 entries and there are five different columns (named title, description, category, language and actor_name). All values in the database have a value (non-null) and the values and dtypes are all str (strings). You can also see the datasets size (10.4kb)

Show column statistics on df2 using the describe method.

In [62]:
df2.describe()

,title,description,category,language,actor_name
count,262,262,262,262,262
unique,46,46,15,1,145
top,ARABIA DOGMA,A Touching Epistle of a Madman And a Mad Cow w...,Horror,English,OPRAH KILMER
freq,12,12,55,262,8


Looking at the describe results, what are the useful insights you can take away? What summarizations are likely not so useful?

The first line just tell me the total entries per column (count). You can see the unique, top and freq for every column. For example, for title there are 46 unique values, the top title is ARABIA.. and the title is freq 12 times in the database.

Create a new variable df3. Use the read_csv function to read the contents of northwind_example.csv into df3. (Note again that you will need to save the CSV file to the same folder where Jupyter Notebook is running for this exercise.) Call df3 to review the result.

In [63]:
df3 = pd.read_csv("northwind_example.csv")
df3

,OrderID,ProductName,UnitPrice,Quantity
0,10248,Queso Cabrales,14.0,12
1,10248,Singaporean Hokkien Fried Mee,9.8,10
2,10248,Mozzarella di Giovanni,34.8,5
3,10249,Manjimup Dried Apples,42.4,40
4,10249,Tofu,18.6,9
...,...,...,...,...
495,10435,Mozzarella di Giovanni,27.8,10
496,10436,Gnocchi di nonna Alice,30.4,40
497,10436,Wimmers gute Semmelkndel,26.6,30
498,10436,Rhnbru Klosterbier,6.2,24


Show column statistics on df3 using the describe method.

In [64]:
df3.describe()

,OrderID,UnitPrice,Quantity
count,500.000000,500.000000,500.000000
mean,10341.364000,23.371400,24.330000
std,54.442663,27.334436,18.069702
min,10248.000000,2.000000,1.000000
25%,10294.000000,10.400000,10.000000
50%,10341.000000,15.800000,20.000000
75%,10389.000000,27.800000,32.000000
max,10436.000000,210.800000,120.000000


Looking at the describe results, what are the useful insights you can take away? How are the summarizations here different than for the previous example, and why?

Similar to the output for the sakila_example.csv, but since this database has actual values "describe" exxecutes more results. You see the columns mean, std, min, 25%, 50%, 75% and max. The data is divided into three different columns, OrderID, UnitPrice and Quantity. So this is a quick summary of the max/min UnitPrice and such.

## Coding a DataFrame.

In [67]:
test_data = {
    'ID': [1,2,3,4,5,6,7,8],
    'Name': ['Alice', 'Bob', 'Charlie', 'Dominic', 'Eve', 'Frank', 'Grace', 'Heidi'],
    'Age': [25,np.nan,22,19,31,35,np.nan,18],
    'City': ['NY','LA','Chicago','Houston','Phoenix','Boston','Austin','San Diego'],
    'Score': [88, 92, -1, 65, 78, np.nan, 55, 90]}

In [68]:
df4 = pd.DataFrame(test_data)
df4

,ID,Name,Age,City,Score
0,1,Alice,25.0,NY,88.0
1,2,Bob,NaN,LA,92.0
2,3,Charlie,22.0,Chicago,-1.0
3,4,Dominic,19.0,Houston,65.0
4,5,Eve,31.0,Phoenix,78.0
5,6,Frank,35.0,Boston,NaN
6,7,Grace,NaN,Austin,55.0
7,8,Heidi,18.0,San Diego,90.0


In [69]:
df4.to_csv("test_data.csv", index=False)